### PR TITLE
Test on GitHub Actions

### PR DESCRIPTION
**Pull request by hugovk**
This tests with the same as Travis CI, and I replaced `nightly` with `3.10-dev`, the nearest equivalent.

Example: https://github.com/hugovk/twitter/actions/runs/360319693

They fail for me because:

```
details: {u'errors': [{u'message': u'User is over daily status update limit.', u'code': 185}]}
```

Will try again tomorrow.

### CODE REVIEW

**Comment by RouxRC** [comment]
Yep, too many tests ran today with the account, will need to retry later

**Comment by RouxRC** [comment]
Hey Hugo,
Sorry with my merges and your push, I guess finding a window for the tests was not easy. Can you ask for a rebuild on your github actions tryout?
Then if it's successful, I guess we should prepare a merge by removing the .travis file first and the relative badge in the readme. Do you know if there is a way to keep some kind of badge as well with GH Actions?

**Comment by hugovk** [comment]
Hi, no problem! 

I restarted a build:

https://github.com/hugovk/twitter/actions/runs/363545104

Some passed, some failed. I'll check the failures tomorrow.

Yes, it has a badge, will be something like:

```md
![Test](https://github.com/sixohsix/twitter/workflows/Test/badge.svg)
```

**Comment by RouxRC** [comment]
Looking at the [test account](https://twitter.com/ptttest0001), it seems like all python versions run simultaneously which might be an issue with the API limits, it is possible to set them to run successively instead of in parallel?

**Comment by hugovk** [comment]
Good idea! Yes, it's possible with `jobs.<job_id>.strategy.max-parallel`:

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymax-parallel

Updated, and that helped:

* https://github.com/hugovk/twitter/actions/runs/366124382

---

2.7 and PyPy2 fails due to:

`CoverallsException: Not on TravisCI. You have to provide either repo_token in .coveralls.yml or set the COVERALLS_REPO_TOKEN env var.`

I'll check that (I'm mostly using Codecov on other projects). Maybe using https://github.com/coverallsapp/github-action or https://github.com/AndreMiras/coveralls-python-action instead will help.

**Comment by hugovk** [comment]
Got Coveralls working with https://github.com/AndreMiras/coveralls-python-action:

* https://github.com/hugovk/twitter/actions/runs/366327891
* https://coveralls.io/builds/34999614

**Comment by RouxRC** [comment]
Yeah !  \o/ 
So I guess we're nearly good to merge?
Do you wanna add the badge change and the travis removal within the PR first, and maybe rename the action as "Tests" with a final "s"?

**Comment by hugovk** [comment]
Yes, I think we're good to merge! I've remove `.travis.yml` and updated the badge here.

File: README [hunk 1]
```diff
@@ -1,7 +1,9 @@
 Python Twitter Tools
 ====================
 
-[![Build Status](https://travis-ci.org/sixohsix/twitter.svg)](https://travis-ci.org/sixohsix/twitter)
+![Test]()
```
Comment (RouxRC) [inline README:4]: what is this line for?

**Review by RouxRC — COMMENTED** [review]
There's just one leftover line I'm not sure about

And out of curiosity, could you tell me what do the pip-cache and actions/cache serve for exactly?

Thanks again!

File: README (hunk 1 above)
Comment (hugovk) [reply README:4]: That's a mistake, will remove it.

I'd pasted the new badge boilerplate there, and copied and pasted out the bits we needed, then forgot to tidy up. Thanks for checking!

**Comment by hugovk** [comment]
> And out of curiosity, could you tell me what do the pip-cache and actions/cache serve for exactly?

Sure, they're for caching the pip downloads, so they don't need to be downloaded from PyPI each time.

PyPI costs $800k/month in hosting! (In March 2020, will be higher now):

* https://twitter.com/dstufft/status/1236331765846990848

I believe this is generously donated by Fastly, but it would be good to help reduce it where possible, and CIs will contribute quite a lot.

---

The config is in two parts:

```yml
      - name: Get pip cache dir
        id: pip-cache
        run: |
          echo "::set-output name=dir::$(pip cache dir)"
```

First, we check the pip cache directory by calling `pip cache dir` and saving the output in a variable called `dir`; we can refer back to it in later steps as `steps.pip-cache.outputs.dir`.

Strictly speaking, this is only needed when testing on multiple operating systems, we could hardcode it as we're only testing on Linux.

More info:

* https://github.com/actions/cache/blob/main/examples.md#python---pip

```yml
      - name: Cache
        uses: actions/cache@v2
        with:
          path: ${{ steps.pip-cache.outputs.dir }}
          key:
            ${{ matrix.os }}-${{ matrix.python-version }}-v1-${{ hashFiles('**/setup.py') }}
          restore-keys: |
            ${{ matrix.os }}-${{ matrix.python-version }}-v1-
```

Next, we use the https://github.com/actions/cache action to fetch cache at the start, and potentially save a new cache at the end.

`path` is the pip dir we saved from the first bit.

`key` is an ID generated and used when saving the cache. We put in the OS and Python version, to make sure we don't mix and match caches between other OS and Python versions (we could remove OS, as we test use Linux right now). We also hash the `setup.py` file so a new cache is created when dependencies change. Often dependencies are listed in `setup.py` or `requirements.txt`. Now I check, there's actually none listed in `setup.py` and there is no `requirements.txt`, so this could be removed, although I doubt it makes too much difference. Finally, the `v1` is there as an easy way to invalidate an old cache and force a new one to be created; because GHA doesn't have a UI to delete old caches.

`restore-keys` is a list of substrings to match old cache names, in case there's none matching `key`.

---

So we could reduce those two chunks to something like:

```yml
      - name: Cache
        uses: actions/cache@v2
        with:
          path: ~/.cache/pip
          key:
            ${{ matrix.python-version }}-v1-${{ hashFiles('**/setup.py') }}
          restore-keys: |
            ${{ matrix.python-version }}-v1-
```

File: README (hunk 1 above)
Comment (hugovk) [reply README]: Removed!

**Comment by RouxRC** [comment]
Ha great, so it's a cache across successive action runs, I wouldn't even have thought it was possible and was not understanding how caching would work across different python versions :)
Makes total sense now, thanks a lot for the explanations, very useful!

**Comment by RouxRC** [comment]
And of course we already sent too many tweets for the day, will rerun the action for the badge to display properly tomorrow!

**Comment by RouxRC** [comment]
And now we're finally good! :)
Thx again for this long run :)

**Comment by hugovk** [comment]
Great, happy to help! And thank you for maintaining this library!

**Comment by RouxRC** [comment]
Now we really need to bother @sixohsix one more time and ask for a release, or better rights on pypi ;)

**Pull request merged**